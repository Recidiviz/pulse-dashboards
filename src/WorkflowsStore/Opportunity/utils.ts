// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { ascending } from "d3-array";
import { differenceInDays, differenceInMonths, format } from "date-fns";
import { snakeCase } from "lodash";
import moment from "moment";
import simplur from "simplur";

import { FeatureGateError } from "../../errors";
import { FeatureVariant } from "../../RootStore/types";
import { pluralizeWord } from "../../utils";
import { AllPossibleKeys } from "../../utils/typeUtils";
import { JusticeInvolvedPersonBase } from "../JusticeInvolvedPersonBase";
import { ValidateFunction } from "../subscriptions";
import { optionalFieldToDate } from "../utils";
import {
  OPPORTUNITY_CONFIGS,
  OpportunityHydratedHeader,
  OpportunityType,
} from "./OpportunityConfigs";
import {
  Opportunity,
  OPPORTUNITY_STATUS_RANKED,
  OpportunityCaseNote,
  OpportunityRequirement,
} from "./types";
import {
  COMPLIANT_REPORTING_ALMOST_CRITERIA_RANKED,
  CompliantReportingOpportunity,
} from "./UsTn";

export function rankByReviewStatus(opp: Opportunity): number {
  if (opp instanceof CompliantReportingOpportunity) {
    if (opp.almostEligible) {
      // sort denials to the bottom
      if (opp.reviewStatus === "DENIED") {
        return COMPLIANT_REPORTING_ALMOST_CRITERIA_RANKED.length;
      }

      const ineligibleCriteriaKeys = opp.record?.ineligibleCriteria
        ? (Object.keys(opp.record.ineligibleCriteria) as Array<
            keyof typeof opp.record.ineligibleCriteria
          >)
        : [];

      return Math.min(
        ...opp.validAlmostEligibleKeys.map((key) =>
          COMPLIANT_REPORTING_ALMOST_CRITERIA_RANKED.indexOf(key)
        ),
        ...ineligibleCriteriaKeys.map((key) =>
          COMPLIANT_REPORTING_ALMOST_CRITERIA_RANKED.indexOf(key)
        )
      );
    }
  }
  return OPPORTUNITY_STATUS_RANKED.indexOf(opp.reviewStatus);
}

export function formatNoteDate(date: Date): string {
  return format(date, "MMMM do");
}

export const generateOpportunityInitialHeader = (
  opportunityType: OpportunityType,
  justiceInvolvedPersonTitle: string,
  workflowsSearchFieldTitle: string
): string => {
  const { label, initialHeader } = OPPORTUNITY_CONFIGS[opportunityType];
  return (
    initialHeader ||
    `Search for ${pluralizeWord(
      workflowsSearchFieldTitle
    )} above to review and refer eligible ${pluralizeWord(
      justiceInvolvedPersonTitle
    )} for ${label.toLowerCase()}.`
  );
};

export const generateOpportunityHydratedHeader = (
  opportunityType: OpportunityType,
  count: number
): OpportunityHydratedHeader => {
  return OPPORTUNITY_CONFIGS[opportunityType].hydratedHeader(count);
};

export function sortByReviewStatus(
  opp1: Opportunity,
  opp2: Opportunity
): number {
  // Use the review status on the opportunity to sort.
  // Compliant Reporting has additional conditions to determine the value of the rank.
  const opp1ReviewStatus = rankByReviewStatus(opp1);
  const opp2ReviewStatus = rankByReviewStatus(opp2);

  return ascending(opp1ReviewStatus, opp2ReviewStatus);
}

export function sortByReviewStatusAndEligibilityDate(
  opp1: Opportunity,
  opp2: Opportunity
): number {
  // First, sort by review status
  const rankSort = sortByReviewStatus(opp1, opp2);
  if (rankSort === 0) {
    // If the ranks are equivalent, sort by eligibilityDate
    if (opp1.eligibilityDate && opp2.eligibilityDate) {
      return ascending(opp1.eligibilityDate, opp2.eligibilityDate);
    }
  }
  return rankSort;
}

export function sortByEligibilityDateUndefinedFirst(
  opp1: Opportunity,
  opp2: Opportunity
): number {
  if (!opp1.eligibilityDate) return -1;
  if (!opp2.eligibilityDate) return 1;
  return ascending(opp1.eligibilityDate, opp2.eligibilityDate);
}

export const opportunityToSortFunctionMapping: Record<
  OpportunityType,
  (a: Opportunity, b: Opportunity) => number
> = {
  earlyTermination: sortByReviewStatus,
  compliantReporting: sortByReviewStatus,
  earnedDischarge: sortByReviewStatusAndEligibilityDate,
  LSU: sortByReviewStatusAndEligibilityDate,
  pastFTRD: sortByReviewStatusAndEligibilityDate,
  supervisionLevelDowngrade: sortByReviewStatus,
  usIdCRCResidentWorker: sortByReviewStatus,
  usIdCRCWorkRelease: sortByReviewStatus,
  usIdExpandedCRC: sortByReviewStatus,
  usIdSupervisionLevelDowngrade: sortByReviewStatusAndEligibilityDate,
  usMiSupervisionLevelDowngrade: sortByReviewStatusAndEligibilityDate,
  usMiClassificationReview: sortByReviewStatusAndEligibilityDate,
  usMiEarlyDischarge: sortByReviewStatusAndEligibilityDate,
  usMeSCCP: sortByReviewStatus,
  usMeWorkRelease: sortByReviewStatus,
  usTnExpiration: sortByReviewStatusAndEligibilityDate,
  usTnCustodyLevelDowngrade: sortByReviewStatus,
  usMoRestrictiveHousingStatusHearing: sortByEligibilityDateUndefinedFirst,
  usMeEarlyTermination: sortByReviewStatus,
  usMiMinimumTelephoneReporting: sortByReviewStatus,
  usMiPastFTRD: sortByReviewStatusAndEligibilityDate,
  usMeFurloughRelease: sortByReviewStatus,
  usCaSupervisionLevelDowngrade: sortByReviewStatusAndEligibilityDate,
  usTnAnnualReclassification: sortByReviewStatus,
};

export const transformCaseNotes = (
  caseNotes: Record<string, Record<string, string>[]> | undefined
): Record<string, OpportunityCaseNote[]> => {
  if (!caseNotes) return {};

  return Object.keys(caseNotes).reduce(
    (processedNotes: Record<string, OpportunityCaseNote[]>, section) => {
      return {
        ...processedNotes,
        [section]: caseNotes[section].map((note) => ({
          noteTitle: note.noteTitle,
          noteBody: note.noteBody,
          eventDate: optionalFieldToDate(note.eventDate),
        })),
      };
    },
    {}
  );
};

export const defaultFormValueJoiner = (
  ...items: (string | undefined)[]
): string => items.filter((item) => item && item !== "").join("\n");

export const formatFormValueDateMMDDYYYYY = (date: string | Date): string =>
  moment(date).format("MM/DD/YYYY");

export const displayString = (
  str: string | undefined,
  prefix?: string
): string => {
  const displayText = str || "";
  return !prefix ? displayText : `${prefix} ${displayText}`;
};

export const displayList = (
  lst: string[] | undefined,
  prefix?: string
): string => {
  const hasContents = lst?.length;
  const displayText = hasContents ? lst.join(", ") : "";
  return !prefix || !hasContents ? displayText : `${prefix} ${displayText}`;
};

export const monthsOrDaysRemainingFromToday = (eligibleDate: Date): string => {
  const months = differenceInMonths(eligibleDate, new Date());
  if (months === 0) {
    return simplur`${differenceInDays(eligibleDate, new Date())} more day[|s]`;
  }
  return simplur`${months} more month[|s]`;
};

type CriteriaGroupKey = "eligibleCriteria" | "ineligibleCriteria";

type WithCriteria = Record<
  CriteriaGroupKey,
  Record<string, object | null | undefined>
>;

export type CopyTuple<K extends string> = [K, OpportunityRequirement];

// Copy is defined as an array rather than a record so it can have a well-defined order
export type CriteriaCopy<R extends WithCriteria> = {
  [K in CriteriaGroupKey]: Array<CopyTuple<AllPossibleKeys<R[K]>>>;
};

// Formatters are defined in a separate dict so the function type can depend on the criterion
export type CriteriaFormatters<R extends WithCriteria> = {
  [K in CriteriaGroupKey]?: {
    [C in keyof R[K]]?: Record<
      string,
      (criterion: Required<R[K]>[C], record: R) => string
    >;
  };
};

export function hydrateCriteria<
  R extends WithCriteria,
  C extends CriteriaGroupKey
>(
  record: R | undefined,
  criteriaGroupKey: C,
  criteriaCopy: CriteriaCopy<R>,
  criteriaFormatters: CriteriaFormatters<R> = {}
): OpportunityRequirement[] {
  if (!record) return [];

  return criteriaCopy[criteriaGroupKey]
    .filter(([criterionKey]) => criterionKey in record[criteriaGroupKey])
    .map(([criterionKey, copy]) => {
      const out: OpportunityRequirement = { ...copy };
      const criterion = record[criteriaGroupKey][criterionKey];

      const formatters = [
        ...Object.entries(
          criteriaFormatters[criteriaGroupKey]?.[criterionKey] ?? {}
        ),
        // Auto-generate fallback formatters
        ...Object.entries(criterion ?? {}).map(
          ([k, v]): [string, () => string] => [
            snakeCase(k).toUpperCase(),
            () => String(v),
          ]
        ),
      ];
      formatters.forEach(([name, fmt]) => {
        const placeholder = `$${name}`;
        const formatted = fmt(criterion, record);
        out.text = out.text.replaceAll(placeholder, formatted);
        if (out.tooltip) {
          out.tooltip = out.tooltip.replaceAll(placeholder, formatted);
        }
      });
      // TODO: Should we catch unreplaced placeholders? Might be false positives
      // since the syntax is so minimal
      return out;
    });
}

export function getFeatureVariantValidator<R>(
  person: JusticeInvolvedPersonBase,
  featureVariant: FeatureVariant,
  actualValidator?: ValidateFunction<R>
): ValidateFunction<R> {
  return (record: R): void => {
    if (!person.rootStore.workflowsStore.featureVariants[featureVariant]) {
      throw new FeatureGateError(
        `Opportunity is not enabled for this user without the ${featureVariant} flag.`
      );
    }
    if (actualValidator) actualValidator(record);
  };
}
