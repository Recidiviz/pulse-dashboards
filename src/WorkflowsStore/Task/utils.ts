/*
 * Recidiviz - a data platform for criminal justice reform
 * Copyright (C) 2023 Recidiviz, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 * =============================================================================
 */
// Contact compliance rules taken from:
// https://github.com/Recidiviz/recidiviz-data/blob/main/recidiviz/calculator/pipeline/utils/state_utils/us_id/us_id_supervision_compliance.py
export const US_ID_SUPERVISION_LEVEL_CONTACT_COMPLIANCE: Record<
  string,
  { contacts: number; days: number }
> = {
  MINIMUM: {
    contacts: 1,
    days: 180,
  },
  MEDIUM: {
    contacts: 2,
    days: 90,
  },
  HIGH: {
    contacts: 2,
    days: 30,
  },
};

// Home visit compliance rules taken from:
// https://drive.google.com/file/d/1qx1S0Z8zfNe53YBImW4As-LoeKH0-YEl/view
export const US_ID_SUPERVISION_LEVEL_HOME_VISIT_COMPLIANCE: Record<
  string,
  { contacts: number; days: number }
> = {
  MINIMUM: {
    contacts: 1,
    days: 365,
  },
  MEDIUM: {
    contacts: 1,
    days: 180,
  },
  HIGH: {
    contacts: 1,
    days: 90,
  },
};
