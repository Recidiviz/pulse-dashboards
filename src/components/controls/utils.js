// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import map from "lodash/fp/map";
import filter from "lodash/fp/filter";

export const excludeOption = (options, optionToExclude) =>
  options.filter((option) => optionToExclude.value !== option.value);

/**
 * Flats grouped options that have value to flat list
 ([
 { value: "All", label: "ALL" },
 { value: "REVOCATION", label: "Revocation" },
 {
    label: "SCI",
    options: [
     { value: "SCI_6", label: "SCI 6 months" },
     { label: "SCI 12 months" },
    ],
    },
 { value: "DA_DETOX", label: "D&A Detox" },
 { label: "Mental Health" },
 ], { value: "All", label: "ALL" })  =>
 [
 { value: "All", label: "ALL" },
 { value: "REVOCATION", label: "Revocation" },
 { value: "SCI_6", label: "SCI 6 months" },
 { value: "DA_DETOX", label: "D&A Detox" },
 ]
 * @param options
 * @returns {*}
 */
export const flatOptions = (options) =>
  options.reduce(
    (acc, option) => [
      ...acc,
      ...(option.value ? [option] : []),
      ...(option.options ? filter("value", option.options) : []),
    ],
    []
  );

export const formatSelectOptionValue = (
  allOptions,
  summingOption,
  selectedOptions,
  isShortFormat = true
) => {
  const selectedValues = map("value", selectedOptions);

  // show option label if only one selected
  if (selectedValues.length === 1) {
    const options = excludeOption(flatOptions(allOptions), summingOption);
    const option = options.find((o) => o.value === selectedValues[0]);
    return option ? option.label : "";
  }

  // show group label if all options in the only one group selected
  const selectedGroups = allOptions
    .filter((o) => o.options)
    .filter((group) =>
      group.options.every((o) => selectedValues.includes(o.value))
    );
  if (
    selectedGroups.length === 1 &&
    selectedGroups[0].options.length === selectedValues.length
  ) {
    return `${selectedGroups[0].label} - ${selectedGroups[0].allSelectedLabel}`;
  }

  if (isShortFormat) {
    return `${selectedOptions.length} Items`;
  }

  const groupOptions = excludeOption(
    flatOptions(selectedGroups),
    summingOption
  );
  const optionLabels = selectedOptions
    .filter(
      (option) =>
        !groupOptions.find((groupOption) => groupOption.value === option.value)
    )
    .map((option) => option.label);
  const groupLabels = map(
    (group) => `${group.label} - ${group.allSelectedLabel}`,
    selectedGroups
  );

  return optionLabels.concat(groupLabels).join(", ");
};

export const getNewOptions = (
  allOptions,
  summingOption,
  selectedOptions = []
) => {
  const options = excludeOption(flatOptions(allOptions), summingOption);
  const selectedValues = map("value", selectedOptions);

  const isNoOptionsSelected = selectedValues.length === 0;

  const isSummingOptionSelected =
    selectedValues.length > 1 &&
    selectedValues[selectedValues.length - 1] === summingOption.value;

  const isAllOptionsSelected = options.every((o) =>
    selectedValues.includes(o.value)
  );

  if (isNoOptionsSelected || isSummingOptionSelected || isAllOptionsSelected) {
    return [summingOption];
  }

  if (selectedValues.includes(summingOption.value)) {
    return selectedOptions.filter((o) => o.value !== summingOption.value);
  }

  return selectedOptions;
};
