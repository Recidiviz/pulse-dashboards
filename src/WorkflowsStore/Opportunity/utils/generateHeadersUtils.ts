// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2023 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { pluralizeWord } from "../../../utils";
import {
  OPPORTUNITY_CONFIGS,
  OpportunityHydratedHeader,
  OpportunityType,
} from "..";

export const generateOpportunityInitialHeader = (
  opportunityType: OpportunityType,
  justiceInvolvedPersonTitle: string,
  workflowsSearchFieldTitle: string
): string => {
  const { label, initialHeader } = OPPORTUNITY_CONFIGS[opportunityType];
  return (
    initialHeader ||
    `Search for ${pluralizeWord(
      workflowsSearchFieldTitle
    )} above to review and refer eligible ${pluralizeWord(
      justiceInvolvedPersonTitle
    )} for ${label.toLowerCase()}.`
  );
};

export type CountFormatter = [
  count: number,
  formatter: (quantity: number) => string | number
];
/** Formatter for opportunity headers */
const hydratedHeaderFormatter = (quantity: number): string | number => {
  return quantity === 0 ? "Some" : quantity;
};

export const generateOpportunityHydratedHeader = (
  opportunityType: OpportunityType,
  count: number
): OpportunityHydratedHeader => {
  return OPPORTUNITY_CONFIGS[opportunityType].hydratedHeader([
    count,
    hydratedHeaderFormatter,
  ]);
};
