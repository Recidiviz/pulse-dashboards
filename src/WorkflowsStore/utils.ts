// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import assertNever from "assert-never";
import {
  add,
  addDays,
  differenceInDays,
  endOfToday,
  getMonth,
  getYear,
  parseISO,
} from "date-fns";
import { Timestamp } from "firebase/firestore";

import { StaffFilter, SystemId } from "../core/models/types";
import { WORKFLOWS_SYSTEM_ID_TO_PAGE, WorkflowsPage } from "../core/views";
import {
  AutoSnoozeUpdate,
  CombinedUserRecord,
  ManualSnoozeUpdate,
  StaffRecord,
} from "../FirestoreStore/types";
import { isDemoMode } from "../utils/isDemoMode";
import {
  INCARCERATION_OPPORTUNITY_TYPES,
  IncarcerationOpportunityType,
  OpportunityType,
  SUPERVISION_OPPORTUNITY_TYPES,
  SupervisionOpportunityType,
} from "./Opportunity/types";

/**
 * Returns a string of the month and year formatted as "MM_YYYY"
 */
export function getMonthYearFromDate(date: Date): string {
  const month = getMonth(date) + 1;
  return month < 10
    ? `0${month}_${getYear(date)}`
    : `${month}_${getYear(date)}`;
}

export function formatSupervisionType(supervisionType: string): string {
  return supervisionType.replace("_", " ");
}

export function dateToTimestamp(isodate: string): Timestamp {
  return new Timestamp(new Date(isodate).getTime() / 1000, 0);
}

export function validatePhoneNumber(phoneNumber: string | undefined): boolean {
  if (!phoneNumber) return false;
  const phoneNumberRegex = /^\d{10}$/;
  return phoneNumberRegex.test(clearPhoneNumberFormatting(phoneNumber));
}

export function clearPhoneNumberFormatting(phoneNumber: string) {
  return phoneNumber.replace(/\D/g, "");
}

export function formatPhoneNumber(phoneNumber: string): string {
  const digitsOnly = clearPhoneNumberFormatting(phoneNumber);
  const formattedPhoneNumber = digitsOnly.replace(
    /(\d{3})(\d{3})(\d{4})/,
    "($1) $2-$3"
  );

  return formattedPhoneNumber;
}

export function formatFacilityHousingUnit(
  facilityId: string | undefined,
  unitId: string | undefined
): string {
  const formattedFacilityHousingUnit = `${facilityId ?? ""}${
    facilityId && unitId ? "/" : ""
  }${unitId ?? ""}`;
  return formattedFacilityHousingUnit;
}

// dates in demo fixtures will be shifted relative to this date
const DEMO_TIMESTAMP = parseISO("2021-12-16");

/**
 * Shifts a given date forward by the difference between the current date
 * and the static "demo date" used for fixture data, bringing the dates
 * in demo fixtures up to date relative to today.
 */
export function shiftDemoDate(storedDate: Date): Date {
  const offsetDays = differenceInDays(new Date(), DEMO_TIMESTAMP);
  return addDays(storedDate, offsetDays);
}

/**
 * Given a raw field from Firestore, converts it to a Date.
 * When Demo Mode is active, it also applies a time shift so that
 * the date from demo fixture data is relevant to the current date.
 */
export function fieldToDate(field: Timestamp | string): Date {
  let result: Date;
  if (typeof field === "string") {
    result = parseISO(field);
  } else {
    result = field.toDate();
  }
  if (isDemoMode()) {
    result = shiftDemoDate(result);
  }

  return result;
}

export function optionalFieldToDate(
  field?: Timestamp | string
): Date | undefined {
  if (field) return fieldToDate(field);
}

export function fieldToDateArray(field: Timestamp[] | string[]): Date[] {
  return field.map((d: Timestamp | string) => fieldToDate(d));
}

export function optionalFieldToDateArray(
  field?: Timestamp[] | string[]
): Date[] | undefined {
  if (field) return fieldToDateArray(field);
}

export function fractionalDateBetweenTwoDates(
  dateLeft: Date | undefined,
  dateRight: Date | undefined,
  fractionalPortion: number
): Date | undefined {
  if (dateLeft && dateRight) {
    const differenceInTime = dateRight.getTime() - dateLeft.getTime();
    const fractionalTime = differenceInTime * fractionalPortion;
    return new Date(dateLeft.getTime() + fractionalTime);
  }
}

export function staffNameComparator(a: StaffRecord, b: StaffRecord): number {
  const alphabetAndSpacesOnlyRegex = /[^a-zA-Z\s]+/g;

  const surnameA = String(a.surname)
    .toLowerCase()
    .replace(alphabetAndSpacesOnlyRegex, "");
  const givenNamesA = String(a.givenNames)
    .toLowerCase()
    .replace(alphabetAndSpacesOnlyRegex, "");

  const surnameB = String(b.surname)
    .toLowerCase()
    .replace(alphabetAndSpacesOnlyRegex, "");
  const givenNamesB = String(b.givenNames)
    .toLowerCase()
    .replace(alphabetAndSpacesOnlyRegex, "");

  if (surnameA > surnameB) return 1;
  if (surnameA < surnameB) return -1;
  if (surnameA === surnameB) {
    return String(givenNamesA).localeCompare(givenNamesB);
  }
  return 0;
}

export const OTHER_KEY = "Other";

type Entries<T> = {
  [K in keyof T]: [K, T[K]];
}[keyof T][];
export const getEntries = <T extends object>(obj: T) =>
  Object.entries(obj) as Entries<T>;

export function getJusticeInvolvedPersonTitle(
  systemId: SystemId | undefined
): string {
  switch (systemId) {
    case "INCARCERATION":
      return "resident";
    case "SUPERVISION":
      return "client";
    case "ALL":
    case undefined:
      return "person";
    default:
      assertNever(systemId);
  }
}

export function getSystemIdFromOpportunityType(
  opportunityType: OpportunityType
): SystemId {
  if (
    SUPERVISION_OPPORTUNITY_TYPES.includes(
      opportunityType as SupervisionOpportunityType
    )
  ) {
    return "SUPERVISION";
  }

  if (
    INCARCERATION_OPPORTUNITY_TYPES.includes(
      opportunityType as IncarcerationOpportunityType
    )
  ) {
    return "INCARCERATION";
  }

  throw new Error(`Unexpected OpportunityType: ${opportunityType}.`);
}

export function getSystemIdFromPage(page: WorkflowsPage): SystemId | undefined {
  if (WORKFLOWS_SYSTEM_ID_TO_PAGE.INCARCERATION.includes(page)) {
    return "INCARCERATION";
  }
  if (WORKFLOWS_SYSTEM_ID_TO_PAGE.SUPERVISION.includes(page)) {
    return "SUPERVISION";
  }
  if (WORKFLOWS_SYSTEM_ID_TO_PAGE.ALL.includes(page)) {
    return "ALL";
  }
}

export function filterByUserDistrict(
  user: CombinedUserRecord
): StaffFilter | undefined {
  let filterValues = user.updates?.overrideDistrictIds;

  if (!filterValues?.length) {
    if (user.info.district) {
      filterValues = [user.info.district];
    } else {
      return undefined;
    }
  }

  return {
    filterField: "district",
    filterValues,
  };
}

export function usCaFilterByRoleSubtype(
  user: CombinedUserRecord
): StaffFilter | undefined {
  // If we don't have a staff record for the user and we've listed them as leadership or they have
  // no role (for recidiviz + offline users), let them search for anyone.
  if (
    !user.info.roleSubtype &&
    (!user.info.role || user.info.role === "leadership_role")
  ) {
    return undefined;
  }

  // At this point, assume they're a line staff or supervisor and we need to restrict to searching
  // for themselves or their district.

  // Regular Parole Agents only get access to their own caseload If we don't have a district for a
  // user and we didn't list them as leadership, restrict them to their own caseload as well.
  if (
    user.info.roleSubtype === "SUPERVISION_OFFICER" ||
    (!user.info.district && !user.updates?.overrideDistrictIds)
  ) {
    return {
      // Use email instead of id because users might not have a staff record or an id set in the
      // admin panel
      filterField: "email",
      filterValues: [user.info.email],
    };
  }

  // Parole Agent supervisors only get access to their own unit (district in our schema)
  return filterByUserDistrict(user);
}

/* Returns the snooze until date from either the auto or manual snooze updates. */
export function getSnoozeUntilDate({
  snoozeUntil,
  snoozeForDays,
  snoozedOn,
}: Partial<AutoSnoozeUpdate | ManualSnoozeUpdate>): Date | undefined {
  if (snoozeUntil) return parseISO(snoozeUntil);
  if (snoozedOn && snoozeForDays)
    return add(parseISO(snoozedOn), { days: snoozeForDays });
}

export function snoozeUntilDateInTheFuture(snoozeUntilDate: Date) {
  return snoozeUntilDate > endOfToday();
}
