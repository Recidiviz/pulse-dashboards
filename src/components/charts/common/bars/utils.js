// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2020 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import groupBy from "lodash/fp/groupBy";
import map from "lodash/fp/map";
import pipe from "lodash/fp/pipe";
import reduce from "lodash/fp/reduce";
import sumBy from "lodash/fp/sumBy";
import toInteger from "lodash/fp/toInteger";
import values from "lodash/fp/values";
import upperCase from "lodash/fp/upperCase";

import { configureDownloadButtons } from "../../../../assets/scripts/utils/downloads";

/**
 * Casts arguments to integer and sum it.
 */
export const sum = (a, b) => toInteger(a) + toInteger(b);

/**
 * Internal function for concatinating arrays at the time of merge.
 * Links: https://lodash.com/docs/4.17.15#mergeWith
 */
export const mergeAllResolver = (objValue, srcValue) =>
  Array.isArray(objValue) ? objValue.concat(srcValue) : [objValue, srcValue];

/**
 * Checks if officer has valid and not empty name.
 */
export const isValidOfficer = (offices) => ({
  officer_external_id: officerIDRaw,
  district: officeId,
}) => {
  const officeName = offices[toInteger(officeId)];
  return officeName && officerIDRaw !== "OFFICER_UNKNOWN";
};

export const isValidOffice = (visibleOffices) => ({ district: officeId }) => {
  if (visibleOffices.length === 1 && visibleOffices[0] === "all") {
    return true;
  }
  return visibleOffices.includes(officeId);
};

/**
 * Groups dataset by month/year and sum all bar counters.
 */
export const groupByMonth = (barKeys) => (dataset) =>
  pipe(
    groupBy((item) => `${item.year}-${item.month}`),
    values,
    map((data) => ({
      year: data[0].year,
      month: data[0].month,
      ...reduce(
        (acc, barKey) => ({
          ...acc,
          [barKey]: sumBy((o) => toInteger(o[barKey]), data),
        }),
        {},
        barKeys
      ),
    }))
  )(dataset);

export function configureDownloads(
  chartId,
  chartLabels,
  countsByType,
  visibleOffices,
  exportLabel,
  bars,
  toggles
) {
  const exportedStructureCallback = () => ({
    office: visibleOffices.join(", "),
    metric: exportLabel,
    series: [],
  });

  const downloadableDataFormat = bars.map((bar) => ({
    label: bar.label,
    data: countsByType[bar.key],
  }));

  const chartTitle = upperCase(exportLabel);

  const convertValuesToNumbers = false;

  configureDownloadButtons(
    chartId,
    chartTitle,
    downloadableDataFormat,
    chartLabels,
    document.getElementById(chartId),
    exportedStructureCallback,
    toggles,
    convertValuesToNumbers
  );
}

export const isOfficerIdsHidden = (offices) =>
  (offices.length === 1 && offices[0] === "all") || offices.length > 3;
