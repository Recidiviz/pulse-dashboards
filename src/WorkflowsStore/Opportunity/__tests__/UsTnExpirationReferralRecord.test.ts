// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2022 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import { identity } from "lodash";

import { fieldToDate, OpportunityValidationError } from "../../utils";
import {
  getValidator,
  transformReferral,
} from "../UsTnExpirationReferralRecord";

const usTnExpirationRecordRaw = {
  stateCode: "US_XX",
  externalId: "abc123",
  criteria: {
    supervisionPastFullTermCompletionDate: {
      eligibleDate: "2022-03-03",
    },
  },
};

const mockClient = {
  expirationDate: fieldToDate("2022-03-03"),
  rootStore: {
    workflowsStore: {
      formatSupervisionLevel: identity,
    },
  },
};

test("transform function", () => {
  expect(transformReferral(usTnExpirationRecordRaw)).toMatchSnapshot();
});

test("record validates", () => {
  const validator = getValidator(mockClient as any);
  expect(() =>
    validator(transformReferral(usTnExpirationRecordRaw))
  ).not.toThrow(OpportunityValidationError);
});

test("record does not validate", () => {
  const validator = getValidator({
    ...mockClient,
    expirationDate: fieldToDate("2022-04-04"),
  } as any);
  expect(() => validator(transformReferral(usTnExpirationRecordRaw))).toThrow(
    OpportunityValidationError
  );
});
