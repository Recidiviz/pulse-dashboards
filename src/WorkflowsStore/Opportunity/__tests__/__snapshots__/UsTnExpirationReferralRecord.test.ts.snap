// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transform function 1`] = `
Object {
  "caseNotes": Object {},
  "criteria": Object {
    "supervisionPastFullTermCompletionDateOrUpcoming60Day": Object {
      "eligibleDate": 2022-03-03T00:00:00.000Z,
    },
    "usTnNoZeroToleranceCodesSpans": Object {
      "zeroToleranceCodeDates": undefined,
    },
    "usTnNotOnLifeSentenceOrLifetimeSupervision": Object {
      "lifetimeFlag": false,
    },
  },
  "externalId": "abc123",
  "stateCode": "US_XX",
}
`;
